src/Common/Core/Impl/Threading/BinaryAsyncLock.cs

```diff
@@ -0,0 +1,30 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Common.Core.Threading {
+    /// <summary>
+    /// BinaryAsyncLock is a helper primitive that can be used instead of SemaphoreSlim.WaitAsync + double-checked locking
+    /// After BinaryAsyncLock is created or reset, the first caller of <see cref="WaitAsync"/> will immediately get <see langword="false" />
+    /// All other callers will wait until <see cref="Release"/> is called and then will get <see langword="true" />
+    /// </summary>
+    public class BinaryAsyncLock {
```

Comment:
So you suggest to define some kind of `BinaryAsyncLockWaitResult` which will be comparable to `bool`:
```{csharp}
using (var areLoaded = await _availableAndInstalledLock.WaitAsync()) {
    if (!areLoaded) {
        IsLoading = true;
        await LoadInstalledAndAvailablePackagesAsync();
        IsLoading = false;
    }
}
```